### PR TITLE
Do not wait for advancement of checkpoint in recovery

### LINKED ISSUE #38949: [CI] IndexFollowingIT.testFollowIndex failure
This test has fail with an assertion error. In the logs what looks relevant Is the following:

```
  1> [2019-02-15T08:37:44,302][INFO ][o.e.c.m.MetaDataMappingService] [leaderm2] [index1/zhspcSQ3Su-DfqnT8FirZg] create_mapping [doc]
  1> [2019-02-15T08:37:44,327][INFO ][o.e.c.m.MetaDataMappingService] [follower1] [index2/48YajD45TdeuzdasRRO35A] create_mapping [doc]
  1> [2019-02-15T08:37:45,353][INFO ][o.e.c.m.MetaDataIndexStateService] [follower1] closing indices [index2/48YajD45TdeuzdasRRO35A]
  1> [2019-02-15T08:37:45,368][INFO ][o.e.c.m.MetaDataIndexStateService] [follower1] completed closing of indices [index2]
  1> [2019-02-15T08:37:45,385][WARN ][o.e.x.c.a.b.TransportBulkShardOperationsAction] [follower1] unexpected error during the primary phase for action [indices:data/write/bulk_shard_operations[s]], request [BulkShardOperationsRequest{historyUUID=N2HT1imJT2-IDpPZ0D6Mog, operations=1, maxSeqNoUpdates=-1, shardId=[index2][0], timeout=1m, index='index2', waitForActiveShards=DEFAULT}]
  1> org.elasticsearch.indices.IndexClosedException: closed
  1> 	at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:215) ~[elasticsearch-7.1.0-SNAPSHOT.jar:7.1.0-SNAPSHOT]
```

It does not reproduce locally:

```
./gradlew :x-pack:plugin:ccr:internalClusterTest \
  -Dtests.seed=37D22C55D4DB9A44 \
  -Dtests.class=org.elasticsearch.xpack.ccr.IndexFollowingIT \
  -Dtests.method="testFollowIndex" \
  -Dtests.security.manager=true \
  -Dtests.locale=es-CU \
  -Dtests.timezone=America/Argentina/Tucuman \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

Console log: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+intake/122/console

### LINKED ISSUE #39000: Replica recovery on follower can fail after recovery from remote
Today during peer recovery, we replay operations from the translog of the primary to the replica. This is to build a history of operations on the replica, in case that it would become a primary. Unfortunately, this is currently fatal for peer recovery of a follower shard after the primary shard has recovered remotely.

With recovery from remote, we copy over the index files and there is no translog replay phase. Assume the simple case of a newly created follower doing its initial recovery from the leader. If the leader does a flush before the follower initiates recovery from remote, after recovery from remote the follower will be fully caught up all primary shards of the follower will have empty translogs. When a replica shard of the follower attempts to recover from the primary shard of the follower, we want to replay translog from the local checkpoint of the commit, to bake a history of operations for it. Since the primary shards of the follower have empty translogs, this replay can not happen and recovery will fail.

Immediately I see two possibilities:
 - `InternalEngine#readHistoryOperations` could serve operations from the Lucene history instead of the translog; however, this interplays with the need for shard history retention leases to interplay with recovery to ensure that these operations exist
 - the primary shards of the follower could build their own translog from Lucene history after remotely recovering files

The purpose of this issue is to tracker this as a blocker for all versions of CCR that support recovery from remote, and to initiate discussion on these options (and others).

Relates #38633

### PR DESCRIPTION
**Pull request by dnhatn**
With this change, we won't wait for the local checkpoint to advance to the max_seq_no before starting phase2 of peer-recovery. We also remove the sequence number range check in peer-recovery. We can safely do these thank to Yannick's finding.

The replication group to be used is currently sampled after indexing into the primary (see `ReplicationOperation` class). This means that when initiating tracking of a new replica, we have to consider the following two cases:

* there are operations for which the replication group has not been sampled yet. As we initiated the new replica as tracking, we know that those operations will be replicated to the new replica and follow the typical replication group semantics (e.g. marked as stale when unavailable).

* there are operations for which the replication group has already been sampled. These operations will not be sent to the new replica. However, we know that those operations are already indexed into Lucene and the translog on the primary, as the sampling is happening after that. This means that by taking a snapshot of Lucene or the translog, we will be getting those ops as well. What we cannot guarantee anymore is that all ops up to `endingSeqNo` are available in the snapshot (i.e. also see comment in `RecoverySourceHandler` saying `We need to wait for all operations up to the current max to complete, otherwise we can not guarantee that all operations in the required range will be available for replaying from the translog of the source.`). This is not needed, though, as we can no longer guarantee that max seq no == local checkpoint.

Relates #39000 
Closes #38949

### CODE REVIEW

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java [hunk 1]
```diff
@@ -57,240 +73,250 @@
...
+        try (ReplicationGroup leaderGroup = createLeaderGroup(randomInt(2))) {
             leaderGroup.startAll();
-            int docCount = leaderGroup.appendDocs(randomInt(64));
-            leaderGroup.assertAllEqual(docCount);
-            followerGroup.startAll();
-            ShardFollowNodeTask shardFollowTask = createShardFollowTask(leaderGroup, followerGroup);
-            final SeqNoStats leaderSeqNoStats = leaderGroup.getPrimary().seqNoStats();
-            final SeqNoStats followerSeqNoStats = followerGroup.getPrimary().seqNoStats();
-            shardFollowTask.start(
+            try (ReplicationGroup followerGroup = createFollowGroup(leaderGroup, randomInt(2))) {
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java]: I can open a separate PR to add a remote-recovery simulation to ShardFollowTaskReplicationTests.

Comment (dnhatn) [reply x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java]: I opened #39007. You only have to look at `testRemoteRecoveryWithGapsInIndexCommit` in this file.

**Comment by dnhatn** [comment]
@bleskes Would you please take a look? Thank you!

**Review by DaveCTurner — COMMENTED** [review]
As per https://github.com/elastic/elasticsearch/issues/39000#issuecomment-465007326 there doesn't seem to be a way for the primary to fail to replicate every operation (thanks in part to the primary/replica resync work) so it sounds like we don't need this check. But we do need tests to ensure that @ywelsch's reasoning doesn't break down in future.

**Comment by dnhatn** [comment]
Closes in favor of https://github.com/elastic/elasticsearch/pull/39153.

**Comment by dnhatn** [comment]
I missed Yannick's comment. Thanks for ping @DaveCTurner. I have updated this PR to remove the range check and the wait for the advancement of the local checkpoint in the recovery source.

**Comment by dnhatn** [comment]
> But we do need tests to ensure that @ywelsch's reasoning doesn't break down in future.

@DaveCTurner I moved the newly added test to #39153.

**Review by jasontedor — APPROVED** [review]
LGTM. Note that you can remove `IndexShard#waitForOpsToComplete` now.

**Comment by dnhatn** [comment]
> Note that you can remove `IndexShard#waitForOpsToComplete` now.

Yeah, I pushed https://github.com/elastic/elasticsearch/pull/39006/commits/ebcebd761d18689849e5a4e0a1743c6bc1952ca8 to remove waitForOpsToComplete in IndexShard and Engine.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Review by ywelsch — APPROVED** [review]
LGTM. Thanks @dnhatn

**Comment by dnhatn** [comment]
@DaveCTurner @jasontedor @ywelsch Thanks so much for reviewing.

**Pull request merged**